_test/test/build_integration_test.dart

```diff
@@ -96,5 +96,17 @@ void main() {
       final nextBuild = await runBuild(trailingArgs: ['--fail-on-severe']);
       expect(nextBuild.exitCode, 0);
     });
+
+    test(
+        'Restores previously deleted outputs if they are not deleted in subsequent builds',
```

Comment:
I went ahead and wrote a regression test but my guess is @jakemac53  won't like it 😝 

I'm also not covering the case where multiple post process builders want to delete the same asset with any test.

We might need to bite the bullet and start writing tests with test_descriptor so we can have isolated little sample packages...

Reply:
Note sure what's wrong with the test. Looks fine to me :/

Reply:
It has some of the same downsides as https://github.com/dart-lang/build/pull/1398#discussion_r185284625 though maybe not quite as drastic

The main thing I don't like about it is having to build with Dart2Js again which slows down the tests.

The test would be faster, and would be a more direct expression of what we're covering, if we make some builders that explicitly do or don't delete their outputs and change only that. The Dart source cleanup builder does that - but it's also tied to other details.